eZ/Publish/Core/Repository/UserService.php

```diff
@@ -731,42 +688,55 @@ public function updateUser(APIUser $user, UserUpdateStruct $userUpdateStruct): A
-            }
+        $this->executeUserUpdate($loadedUser, $userUpdateStruct);
 
-            if ($userUpdateStruct->contentMetadataUpdateStruct !== null) {
-                $contentService->updateContentMetadata(
-                    $publishedContent->getVersionInfo()->getContentInfo(),
-                    $userUpdateStruct->contentMetadataUpdateStruct
-                );
-            }
+        return $this->loadUser($loadedUser->id);
+    }
 
-            // User\Handler::update call is currently used to clear cache only
-            $this->userHandler->update(
-                new SPIUser(
-                    [
-                        'id' => $loadedUser->id,
-                        'login' => $loadedUser->login,
-                        'email' => $userUpdateStruct->email ?: $loadedUser->email,
-                    ]
-                )
-            );
+    public function updateUserPassword(APIUser $user, UserUpdateStruct $userUpdateStruct): APIUser
+    {
+        $loadedUser = $this->loadUser($user->id);
 
-            $this->repository->commit();
-        } catch (Exception $e) {
-            $this->repository->rollback();
-            throw $e;
+        $contentService = $this->repository->getContentService();
+
+        if (!$this->permissionResolver->canUser('content', 'edit', $loadedUser)
+            && !$this->permissionResolver->canUser('user', 'password', $loadedUser)
+        ) {
+            throw new UnauthorizedException('user', 'password');
+        }
+
+        if (empty($userUpdateStruct->password)) {
+            throw new \InvalidArgumentException('User update struct does not contain password');
         }
 
+        $userFieldDefinition = $this->getUserFieldDefinition($loadedUser->getContentType());
+        if ($userFieldDefinition === null) {
+            throw new MissingUserFieldTypeException($loadedUser->getContentType());
+        }
+
+        $userUpdateStruct->contentUpdateStruct = $contentService->newContentUpdateStruct();
+        $userUpdateStruct->contentUpdateStruct->setField(
+            $userFieldDefinition->identifier,
+            new UserValue([
+                'contentId' => $loadedUser->id,
+                'hasStoredLogin' => true,
+                'login' => $loadedUser->login,
+                'email' => $loadedUser->email,
+                'plainPassword' => $userUpdateStruct->password,
+                'enabled' => $loadedUser->enabled,
+                'maxLogin' => $loadedUser->maxLogin,
+                'passwordHashType' => $user->hashAlgorithm,
+                'passwordHash' => $user->passwordHash,
+            ])
+        );
+
+        $updateUserCallable = function () use ($loadedUser, $userUpdateStruct): void {
+            $this->executeUserUpdate($loadedUser, $userUpdateStruct);
+        };
+
+        // Perform update without checking other user permissions
+        $this->repository->sudo($updateUserCallable);
```

Comment:
Do we really need this `sudo`? I am aware of this discussion https://github.com/ezsystems/ezplatform-kernel/pull/117#discussion_r503732769 but still, `sudo` feels like a hack.

What about changes to `owner:self` limitation? Won't it solve the issue and remove the need of `sudo`?

Reply:
As discussed, limitation itself will remain unchanged, but instead of calling `$this->repository->sudo()` user password update will be handled by a dedicated SPI method.

Reply:
Implementation has been changed to omit using `ContentService`. New dependencies have been added to `UserService` as a result and new SPI method has been added.